reasoner/planner/ConjunctionStreamPlan.java

```diff
@@ -0,0 +1,337 @@
+                    String.join(", ", iterate(extendOutputWith).map(Identifier.Variable::toString).toList()),
+                    String.join(", ", iterate(outputVariables).map(Identifier.Variable::toString).toList()),
+                    String.join(" ; ", iterate(subPlans).map(ConjunctionStreamPlan::toString).toList()));
+        }
+    }
+
+    private static class Builder {
+        private final List<Resolvable<?>> resolvables;
+        private final Set<Retrievable> inputBounds;
+        private final Set<Retrievable> outputBounds;
+
+        private final List<Set<Retrievable>> boundsAt;
+
+        private Builder(List<Resolvable<?>> resolvables, Set<Retrievable> inputBounds, Set<Retrievable> outputBounds) {
+            this.resolvables = resolvables;
+            this.inputBounds = inputBounds;
+            this.outputBounds = outputBounds;
+            boundsAt = new ArrayList<>();
+            Set<Retrievable> runningBounds = new HashSet<>(inputBounds);
+            for (Resolvable<?> resolvable : resolvables) {
+                boundsAt.add(new HashSet<>(runningBounds));
+                iterate(resolvable.retrieves()).filter(Retrievable::isName).forEachRemaining(runningBounds::add);
+            }
+        }
+
+        private ConjunctionStreamPlan build() {
+            return buildRecursivePrefix(resolvables, inputBounds, outputBounds);
+        }
+
+        public ConjunctionStreamPlan buildRecursivePrefix(List<Resolvable<?>> subConjunction, Set<Retrievable> availableInputs, Set<Retrievable> requiredOutputs) {
+            if (subConjunction.size() == 1) {
+                VariableSets variableSets = VariableSets.determineVariableSets(list(), subConjunction, availableInputs, requiredOutputs);
+                //  use resolvableOutputs instead of rightOutputs because this node has to do the job of the parent as well - joining the identifiers
+                Set<Retrievable> resolvableOutputs = VariableSets.difference(requiredOutputs, variableSets.extendOutputWith);
+                return new ResolvablePlan(subConjunction.get(0), variableSets.rightInputs, variableSets.extendOutputWith, resolvableOutputs);
+            } else {
+                Pair<List<Resolvable<?>>, List<Resolvable<?>>> divided = divide(subConjunction);
+                VariableSets variableSets = VariableSets.determineVariableSets(divided.first(), divided.second(), availableInputs, requiredOutputs);
+                ConjunctionStreamPlan leftPlan = buildRecursivePrefix(divided.first(), variableSets.leftIdentifiers, variableSets.leftOutputs);
+                ConjunctionStreamPlan rightPlan = buildRecursiveSuffix(divided.second(), variableSets.rightInputs, variableSets.rightOutputs);
+
+                return new CompoundStreamPlan(list(leftPlan, rightPlan), variableSets.identifiers, variableSets.extendOutputWith, requiredOutputs);
+            }
+        }
```

Comment:
Computes the split into the left and right subtrees, and recurses on the left.
The right has a predictable structure so there's nothing to compute.